Garmin Descent Mk2/Mk2i MTP support

dirkhh opened the pull request:
The Descent Mk2/Mk2i no longer appears as a USB storage / FAT device, but instead as an MTP device.
Instead of adding yet another transport definition, I decided to implement this as two alternative methods of accessing the files directly in `garmin.c`.

dirkhh committed:
Garmin: pass the model to the device_open function

We do this for a few other device where we need slightly different behavior,
depending on the specific model.

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

dirkhh committed:
Garmin: add the Descent Mk2/Mk2i

Just like for the Descent Mk1 this uses the internal Garmin model number (as
indicated in the FIT format) as model, not the USB product ID.

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

dirkhh committed:
Garmin: refactor the code to prepare for MTP support

Breaking things into helper function makes it much easier to then support both
regular file operations and MTP.

This includes some mild whitespace cleanups for consistency.

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

dirkhh committed:
Garmin: add MTP support

Create parallel helper functions that use libmtp to walk the file tree on the
device and to then read a specific file from the device into our dc_buffer.
MTP is not a file system, it's an object storage, that just happens to allow
object names and parent/child relationships between objects. As a result we
need to remember those file ids for MTP downloads.

The mtp_get_file_list function is rather complex as it includes both the
initial communication with the device and the code to walk the object tree and
then create the list of file.

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

torvalds commented on src/descriptor.c:
```diff
@@ -406,8 +406,10 @@ static const dc_descriptor_t g_descriptors[] = {
+	/* Garmin -- model numbers as defined in FIT format; USB product id is (0x4000 | model) */
 	{"Garmin", "Descent Mk1", DC_FAMILY_GARMIN, 2859, DC_TRANSPORT_USBSTORAGE, dc_filter_garmin},
+	{"Garmin", "Descent Mk2/Mk2i", DC_FAMILY_GARMIN, 3258, DC_TRANSPORT_USBSTORAGE, dc_filter_garmin},
+	{"Garmin", "Descent Mk2/Mk2i", DC_FAMILY_GARMIN, 3702, DC_TRANSPORT_USBSTORAGE, dc_filter_garmin},
```
Ugh. This means that people see two different entries that look identical. That seems horribly broken.

Plus subsurface will then match on the name when doing the whole "use previously selected dive computer", and just pick the first one. So having two different entries doesn't even work, I think. No?

dirkhh replied:
Ah, right. No, that makes no sense. I'll fix that - I learned about that second model just yesterday and of course don't have access to one (APAC only, not released,yet) so I just added that blind and didn't think it all the way through, I guess.

torvalds commented on src/garmin.c:
```diff
@@ -156,11 +186,23 @@ make_space(struct file_list *files)
+		}
+#else
+		(void)is_mtp;
+#endif
```
Hmm.

I'd just unconditionally add "unsigned int mtp_id" to "struct fit_name" instead. The non-mtp case can just set it to zero and ignore it.

You'd have to make the "name_cmp()" function actually do something like


    static int name_cmp(const void *_a, const void *_b)
    {
        const struct fit_name *a = _a;
        const struct fit_name *b = _b;

        // Sort reverse string ordering (newest first), so use 'b,a'
        return strcmp(b->name, a->name);
    }

and maybe there is something else that depends on "struct fit_name" only being the name, but that would seem to be a cleanup regardless.

dirkhh replied:
ok, I'll look into that

torvalds reviewed (commented):
Small comments for what I think would be improvements, but looks sane to me.

dirkhh commented:
OK, I think I have addressed your comments. Ready to merge as far as I am concerned.

torvalds reviewed (commented):
Looks fine to me, but will double-check when home

torvalds commented:
Ahhah. This thing is somewhat problematic, in that it does that "use_mtp" thing based on the model number of the device you've picked.

Which means that you can't select the Mk2, and then a FIT file that you've downloaded from Garmin onto your local filesystem.

I think the code should probably have a  fallback to the old USBSTORAGE code.

But I don't think this should hold you up from merging it (and merging the subsurface part - they need to go hand-in-hand because updating just libdivecomputer will then cause failures when building subsurface because it now wants the new library)

dirkhh commented:
Ha - that's an interesting aspect that I had not taken into account.
I think the answer is that the Garmin code should assume MTP if there is no path given, but file based if the user passes in a path. I'll take a look at that.

dirkhh committed:
Garmin: support .FIT files from filesystem even for MTP devices

If the caller provides us with a path name, we should not try to connect to the
dive computer via MTP, but instead read the .FIT files from the path provided.

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

torvalds reviewed (approved):
This looks reasonable. Feel free to merge it all up, and I'll do the air integration today.

dirkhh commented:
> This looks reasonable. Feel free to merge it all up, and I'll do the air integration today.

will do. I'm using the holes in my schedule to find time to work on this :-)
